charms/kfp-profile-controller/tests/unit/test_operator.py

```diff
@@ -1,194 +1,223 @@
+    "namespace": "namespace",
+    "port": 1234,
+    "secure": True,
+}
+
+EXPECTED_ENVIRONMENT = {
+    "CONTROLLER_PORT": 80,
+    "DISABLE_ISTIO_SIDECAR": "false",
+    "KFP_DEFAULT_PIPELINE_ROOT": "",
+    "KFP_VERSION": "2.0.0-alpha.7",
+    "METADATA_GRPC_SERVICE_HOST": "mlmd.kubeflow",
+    "METADATA_GRPC_SERVICE_PORT": "8080",
+    "MINIO_ACCESS_KEY": "access-key",
+    "MINIO_HOST": "service",
+    "MINIO_NAMESPACE": "namespace",
+    "MINIO_PORT": 1234,
+    "MINIO_SECRET_KEY": "secret-key",
+    "minio-secret": '{"secret": {"name": ' '"kfp-profile-controller-minio-credentials"}}',
+}
 
-# TODO: Tests missing for dropped/reloaded relations
 
+@pytest.fixture
+def harness() -> Harness:
+    harness = Harness(KfpProfileControllerOperator)
+    return harness
+
+
+@pytest.fixture()
+def mocked_lightkube_client(mocker):
+    """Mocks the Lightkube Client in charm.py, returning a mock instead."""
+    mocked_lightkube_client = MagicMock()
+    mocker.patch("charm.lightkube.Client", return_value=mocked_lightkube_client)
+    yield mocked_lightkube_client
+
+
+# Helpers
+def add_data_to_sdi_relation(
+    harness: Harness,
+    rel_id: int,
+    other: str,
+    data: Optional[dict] = None,
+    supported_versions: str = "- v1",
+) -> None:
+    """Add data to an SDI-backed relation."""
+    if data is None:
+        data = {}
+
+    harness.update_relation_data(
+        rel_id,
+        other,
+        {"_supported_versions": supported_versions, "data": yaml.dump(data)},
+    )
+
+
+def add_sdi_relation_to_harness(
+    harness: Harness, relation_name: str, other_app: str = "other", data: Optional[dict] = None
+) -> dict:
+    """Relates a new app and unit to an sdi-formatted relation.
+
+    Args:
+        harness: the Harness to add a relation to
+        relation_name: the name of the relation
+        other_app: the name of the other app that is relating to our charm
+        data: (optional) the data added to this relation
+
+    Returns dict of:
+    * other (str): The name of the other app
+    * other_unit (str): The name of the other unit
+    * rel_id (int): The relation id
+    * data (dict): The relation data put to the relation
+    """
+    if data is None:
+        data = {}
+
+    other_unit = f"{other_app}/0"
+    rel_id = harness.add_relation(relation_name, other_app)
+
+    harness.add_relation_unit(rel_id, other_unit)
+
+    add_data_to_sdi_relation(harness, rel_id, other_app, data)
+
+    return {
+        "other_app": other_app,
+        "other_unit": other_unit,
+        "rel_id": rel_id,
+        "data": data,
+    }
 
-def test_not_leader(harness):
+
+def test_not_leader(harness, mocked_lightkube_client):
+    """Test when we are not the leader."""
     harness.begin_with_initial_hooks()
-    assert harness.charm.model.unit.status == WaitingStatus("Waiting for leadership")
+    assert harness.charm.model.unit.status == WaitingStatus(
+        "[leadership-gate] Waiting for leadership"
+    )
```

Comment:
I refactored this a bit in [my charm](https://github.com/canonical/kfp-operators/pull/264/files#diff-020e4b99625dc8e2403a768723e645a89483758caf4d651de1a8baa3a4e94308R24-R29) based on your comments today.  Does it feel better?  It feels slightly less coupled to the implementation details of the reusable Component, but maybe not perfect